stdlib/Profile/src/memory.jl

```diff
@@ -0,0 +1,348 @@
+                tag_filter::UInt16 = build_tag_filter())
+    bt_size = something(bt_size, get_memprofile_bt_data_maxlen())
+    alloc_size = something(alloc_size, get_memprofile_alloc_data_maxlen())
+
+    # Sub off to our friend
+    return init(bt_size, alloc_size, tag_filter)
+end
+
+function init(bt_size::Integer, alloc_size::Integer, tag_filter::UInt16)
+    status = ccall(:jl_memprofile_init, Cint, (Csize_t, Csize_t, UInt16), bt_size, alloc_size, tag_filter)
+    if status == -1
+        error("could not allocate space for ", bt_size, " instruction pointers and ", alloc_size, " allocation event structures")
+    end
+end
+
+# This struct must stay in sync with `src/memprofiler.c`
+struct allocation_info_t
+    address::Ptr{Cvoid}
+    T::Ptr{Cvoid}
+    time::Float64
+    allocsz::Csize_t
+    tag::UInt16
+end
+
+# This struct is a combined version of the above `allocation_info_t` with backtrace data
+# and automatic GC marking for the objects that are still open so that they don't get
+# collected while we're modifying them.
+struct AllocationInfo
+    address::UInt
+    T::Any
+    alloc_time::Float64
+    free_time::Float64
+    alloc_stacktrace::Vector{StackFrame}
+    free_stacktrace::Vector{StackFrame}
+    allocsz::UInt64
+    tag::UInt16
+
+    # For an "open" allocation, this is the object that is still alive.  For a "closed"
+    # allocation, this is just `nothing`.
+    object::Any
+end
+
+# If a chunk is "open", we know it's live so actually grab the object and its type
+# The only subtlety is when it's a malloc'd list.
+function update_live_object(a::AllocationInfo)
+    array_addr = ccall(:jl_memprofile_find_malloc_array, Ptr{Cvoid}, (Ptr{Cvoid},), Ptr{Cvoid}(a.address))
+    if array_addr != C_NULL
+        obj = unsafe_pointer_to_objref(array_addr)
+    else
+        # Not doing this right now, because it appears that they can ~~disappear~~
+        obj = nothing # unsafe_pointer_to_objref(Ptr{Cvoid}(a.address))
+    end
+    
+    return AllocationInfo(
+        a.address,
+        a.T,
+        a.alloc_time,
+        a.free_time,
+        a.alloc_stacktrace,
+        a.free_stacktrace,
+        a.allocsz,
+        a.tag,
+        obj
+    )
+end
+
+function Base.show(io::IO, a::AllocationInfo)
+    md = join(tag_lookup(memory_domain_map, a.tag), ", ")
+    at = join(tag_lookup(allocator_map, a.tag), ", ")
+    T = a.T === nothing ? "(?)" : string(a.T)
+    print(io, "AllocationInfo (@$(repr(a.address)): $(a.allocsz) B, domain: ($(md)), allocator: ($(at)), type: $(T))")
+end
+
+"""
+    open_AI(a::allocation_info_t, alloc_stacktrace)
+
+Create an `AllocationInfo` that represents an allocated (and not yet deallocated) chunk
+of memory.  This `AllocationInfo` will later be "closed" to represent the full lifecycle
+of an allocated and then freed piece of memory.  This function has also been known to
+create highly competitive dota agents.
+"""
+function open_AI(a::allocation_info_t, alloc_stacktrace)
```

Comment:
Why not just extend `Base.open`/`close` for this?

Reply:
Because I don't really like that kind of overloading.  It feels like a type pun, there's nothing about this kind of "opening" and "closing" that is similar to `Base.open` or `Base.close`, and so there's not reason for things like docstrings, method tables, etc... to be shared between them.